File: src/utils/geocodingPipeline.ts
/**
 * GEOSPATIAL DATA PIPELINE - ANTIGRAVITY ENGINE
 * Arquitetura Determinística de Geocodificação & Saneamento Cadastral
 * 
 * Regras de Ouro:
 * 1. Sanitização estrita via Regex (isolamento de tipo, nome, número, complemento e CEP).
 * 2. Cascata Hierárquica: Nível 1 (Base Interna) -> Nível 2 (Geocode Predial) -> Nível 3 (Interpolação de Via).
 * 3. Bloqueio Rigoroso de Falsa Precisão: JAMAIS atribuir centróides de bairro, cidade ou estado.
 * 4. Clusterização e Spiderfy para unidades sobrepostas no mesmo condomínio/edifício.
 */

export type StatusGeocodificacao = 
  | 'EXATO_PREDIO'
  | 'GEOCODE_NUMERO'
  | 'INTERPOLACAO_RUA'
  | 'PENDENTE_REVISAO';

export interface ParsedAddress {
  logradouro_tipo: string;
  logradouro_nome: string;
  numero: string | null;
  complemento: string;
  bairro: string;
  cidade: string;
  uf: string;
  cep: string | null;
}

export interface GeocodedRecord {
  id: string;
  endereco_original: string;
  logradouro: string;
  numero: string | null;
  bairro: string;
  cep: string | null;
  lat: number | null;
  lon: number | null;
  status_geocodificacao: StatusGeocodificacao;
  precisa_revisao: boolean;
  detalhe_resolucao?: string;
  cluster_id?: string;
}

// --------------------------------------------------------------------------
// ETAPA 1: Algoritmo Estrito de Parsing e Limpeza (Regex)
// --------------------------------------------------------------------------

export function normalizeStr(str: string | null | undefined): string {
  if (!str) return '';
  return str
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .trim();
}

export function parseAddressStrict(
  rawAddress: string,
  rawNeighborhood: string = '',
  rawCity: string = '',
  rawState: string = '',
  rawCep: string = ''
): ParsedAddress {
  let text = normalizeStr(rawAddress);

  // 1. Extração Estrita de CEP: ^\d{5}-?\d{3}$
  let cep: string | null = null;
  const cepMatch = rawAddress.match(/\b(\d{5})-?(\d{3})\b/) || (rawCep ? rawCep.match(/\b(\d{5})-?(\d{3})\b/) : null);
  if (cepMatch) {
    cep = `${cepMatch[1]}-${cepMatch[2]}`;
  }

  // Remove CEP do corpo do endereço para não poluir nome de rua ou número
  text = text
    .replace(/\b\d{5}-?\d{3}\b/g, '')
    .replace(/\bcep:?\b/i, '')
    .trim();

  // 2. Isolar Prefixo Padronizado (logradouro_tipo)
  const prefixRegex = /^(rua|r\.|r\b|avenida|av\.|av\b|estrada|estr\.|estr\b|travessa|trav\.|trav\b|alameda|al\.|al\b|praca|praça|pca\.|pca\b|rodovia|rod\.|rod\b|largo|lrg\.|lrg\b|beco|bc\.|bc\b|ladeira|lad\.|lad\b)/i;
  const prefixMatch = text.match(prefixRegex);
  let logradouro_tipo = 'Rua';
  let remaining = text;

  if (prefixMatch) {
    const rawP = prefixMatch[1].toLowerCase().replace('.', '');
    if (rawP.startsWith('av')) logradouro_tipo = 'Avenida';
    else if (rawP.startsWith('estr')) logradouro_tipo = 'Estrada';
    else if (rawP.startsWith('tra')) logradouro_tipo = 'Travessa';
    else if (rawP.startsWith('al')) logradouro_tipo = 'Alameda';
    else if (rawP.startsWith('pr') || rawP.startsWith('pc')) logradouro_tipo = 'Praca';
    else if (rawP.startsWith('rod')) logradouro_tipo = 'Rodovia';
    else if (rawP.startsWith('l')) logradouro_tipo = 'Largo';
    else if (rawP.startsWith('bec') || rawP.startsWith('bc')) logradouro_tipo = 'Beco';
    else if (rawP.startsWith('lad')) logradouro_tipo = 'Ladeira';
    else logradouro_tipo = 'Rua';

    remaining = text.slice(prefixMatch[0].length).trim();
  }

  remaining = remaining.replace(/^[-,\.\s]+/, '').trim();

  // 3. Isolar Número Predial Principal e Descartar Complementos
  let numero: string | null = null;
  let complemento = '';
  let logradouro_nome = remaining;

  // Verificação de Sem Número (S/N)
  const snRegex = /\b(s\/n|sem numero|s\/ numero|s\.n\.)\b/i;
  if (snRegex.test(remaining)) {
    numero = null;
    const parts = remaining.split(snRegex);
    logradouro_nome = parts[0];
    complemento = parts.slice(1).join(' ');
  } else {
    // Captura padrão: Rua [Nome], N. [Numero] [Complemento]
    const numRegex = /^(.*?)(?:[,\s]+(?:n[ºo°\.]*|num|numero|n)\s*:?\s*|[,\s]+)(\d{1,6})\b(.*)$/i;
    const numMatch = remaining.match(numRegex);

    if (numMatch) {
      logradouro_nome = numMatch[1];
      numero = numMatch[2];
      complemento = numMatch[3] || '';
    } else {
      // Caso sem número explícito, verificar complementos
      const compRegex = /\b(apto|apt|apartamento|ap|bl|bloco|qd|quadra|lt|lote|cs|casa|fundos|fdo|sobrado|sob|sala|unidade|un|pavimento|pav|andar|edificio|edf|condominio)\b/i;
      const compIndex = remaining.search(compRegex);
      if (compIndex !== -1) {
        logradouro_nome = remaining.slice(0, compIndex);
        complemento = remaining.slice(compIndex);
      }
    }
  }

  // Sanitizar logradouro_nome
  logradouro_nome = logradouro_nome
    .replace(/[,\-\.]+/g, ' ')
    .replace(/\s+/g, ' ')
    .trim()
    .toLowerCase()
    .split(' ')
    .filter(Boolean)
    .map(w => w.charAt(0).toUpperCase() + w.slice(1))
    .join(' ');

  // Sanitizar complemento
  complemento = complemento
    .replace(/^[-,\.\s:]+|[-,\.\s:]+$/g, '')
    .replace(/\s+/g, ' ')
    .toUpperCase();

  // Bairro, Cidade, UF
  const bairro = normalizeStr(rawNeighborhood).toUpperCase().replace(/[^A-Z0-9\s]/g, '').trim();
  const cidade = rawCity ? normalizeStr(rawCity).replace(/[^a-zA-Z0-9\s]/g, '').trim() : 'Rio de Janeiro';
  const uf = rawState ? rawState.toUpperCase().trim() : 'RJ';

  return {
    logradouro_tipo,
    logradouro_nome,
    numero,
    complemento,
    bairro,
    cidade,
    uf,
    cep
  };
}

// --------------------------------------------------------------------------
// ETAPA 2 & 3: Cascata Hierárquica e Bloqueio de Falsa Precisão
// --------------------------------------------------------------------------

export interface InternalGeoCache {
  getExactBuildingCoords?: (street: string, number: string, neighborhood: string) => [number, number] | null;
  getStreetAxisCoords?: (street: string, neighborhood: string, city: string) => [number, number] | null;
}

export async function resolveCoordinatesCascade(
  id: string,
  rawAddress: string,
  rawNeighborhood: string = '',
  rawCity: string = '',
  rawState: string = '',
  rawCep: string = '',
  cache?: InternalGeoCache
): Promise<GeocodedRecord> {
  const parsed = parseAddressStrict(rawAddress, rawNeighborhood, rawCity, rawState, rawCep);
  const fullLogradouro = `${parsed.logradouro_tipo} ${parsed.logradouro_nome}`.trim();

  // NÍVEL 1: Base Própria de Escrituras / IPTU / ITBI (EXATO_PREDIO)
  if (parsed.numero && cache?.getExactBuildingCoords) {
    const internalCoords = cache.getExactBuildingCoords(
      parsed.logradouro_nome.toLowerCase(),
      parsed.numero,
      parsed.bairro.toLowerCase()
    );

    if (internalCoords && internalCoords[0] !== 0 && internalCoords[1] !== 0) {
      return {
        id,
        endereco_original: rawAddress,
        logradouro: fullLogradouro,
        numero: parsed.numero,
        bairro: parsed.bairro,
        cep: parsed.cep,
        lat: Number(internalCoords[0].toFixed(6)),
        lon: Number(internalCoords[1].toFixed(6)),
        status_geocodificacao: 'EXATO_PREDIO',
        precisa_revisao: false,
        detalhe_resolucao: 'Base Oficial de Escrituras/ITBI Municipal'
      };
    }
  }

  // NÍVEL 2 & NÍVEL 3: Geocodificação Estruturada via API Externa (Nominatim)
  try {
    const streetParam = parsed.numero 
      ? `${parsed.numero} ${parsed.logradouro_tipo} ${parsed.logradouro_nome}`
      : `${parsed.logradouro_tipo} ${parsed.logradouro_nome}`;

    const queryParams = new URLSearchParams({
      street: streetParam,
      city: parsed.cidade,
      state: parsed.uf,
      country: 'Brazil',
      format: 'jsonv2',
      addressdetails: '1'
    });

    if (parsed.cep) {
      queryParams.set('postalcode', parsed.cep);
    }

    const apiUrl = `https://nominatim.openstreetmap.org/search?${queryParams.toString()}`;
    
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 2500);

    const response = await fetch(apiUrl, {
      headers: { 'User-Agent': 'MarcusAssessoriaGarimpo/2.0 (Geocoding Engine)' },
      signal: controller.signal
    });
    clearTimeout(timeoutId);

    if (response.ok) {
      const results = await response.json();

      if (Array.isArray(results) && results.length > 0) {
        const top = results[0];
        const lat = parseFloat(top.lat);
        const lon = parseFloat(top.lon);
        const placeRank = top.place_rank || 0;
        const category = top.category || '';
        const type = top.type || '';

        // NÍVEL 2: Nível Predial / Lote / Número (house, building, address, place_rank >= 30)
        if (placeRank >= 30 || type === 'house' || type === 'building' || category === 'building') {
          return {
            id,
            endereco_original: rawAddress,
            logradouro: fullLogradouro,
            numero: parsed.numero,
            bairro: parsed.bairro,
            cep: parsed.cep || top.address?.postcode || null,
            lat: Number(lat.toFixed(6)),
            lon: Number(lon.toFixed(6)),
            status_geocodificacao: 'GEOCODE_NUMERO',
            precisa_revisao: false,
            detalhe_resolucao: `Nominatim Predial (place_rank ${placeRank}, ${type})`
          };
        }

        // NÍVEL 3: Interpolação de Via / Eixo de Rua (highway, residential, road, place_rank 26-28)
        if ((placeRank >= 26 && placeRank <= 28) || category === 'highway') {
          return {
            id,
            endereco_original: rawAddress,
            logradouro: fullLogradouro,
            numero: parsed.numero,
            bairro: parsed.bairro,
            cep: parsed.cep || top.address?.postcode || null,
            lat: Number(lat.toFixed(6)),
            lon: Number(lon.toFixed(6)),
            status_geocodificacao: 'INTERPOLACAO_RUA',
            precisa_revisao: false,
            detalhe_resolucao: `Interpolação no Eixo da Via (place_rank ${placeRank})`
          };
        }
      }
    }
  } catch (err) {
    // Timeout ou erro de rede: continua para cache local
  }

  // Fallback seguro Nível 3: Cache local de eixos de rua conhecidos
  if (cache?.getStreetAxisCoords) {
    const streetAxis = cache.getStreetAxisCoords(
      parsed.logradouro_nome.toLowerCase(),
      parsed.bairro.toLowerCase(),
      parsed.cidade.toLowerCase()
    );

    if (streetAxis && streetAxis[0] !== 0 && streetAxis[1] !== 0) {
      return {
        id,
        endereco_original: rawAddress,
        logradouro: fullLogradouro,
        numero: parsed.numero,
        bairro: parsed.bairro,
        cep: parsed.cep,
        lat: Number(streetAxis[0].toFixed(6)),
        lon: Number(streetAxis[1].toFixed(6)),
        status_geocodificacao: 'INTERPOLACAO_RUA',
        precisa_revisao: false,
        detalhe_resolucao: 'Base Cartográfica Local de Eixo de Vias'
      };
    }
  }

  // ETAPA 3: Bloqueio Rigoroso de Falsa Precisão (Regra de Ouro)
  // Sob nenhuma hipótese atribua coordenadas do centro do bairro ou da cidade!
  return {
    id,
    endereco_original: rawAddress,
    logradouro: fullLogradouro,
    numero: parsed.numero,
    bairro: parsed.bairro,
    cep: parsed.cep,
    lat: null,
    lon: null,
    status_geocodificacao: 'PENDENTE_REVISAO',
    precisa_revisao: true,
    detalhe_resolucao: 'Logradouro não validado com certeza cartográfica predial ou viária'
  };
}

// --------------------------------------------------------------------------
// ETAPA 4: Deduplicação de Coordenadas (Spiderfy) & Estrutura GeoJSON
// --------------------------------------------------------------------------

export interface SpiderfiedLocation {
  id: string;
  originalLat: number;
  originalLon: number;
  displayLat: number;
  displayLon: number;
  isSpiderfied: boolean;
  spiderClusterCount: number;
}

export function applySpiderfyClustering(records: GeocodedRecord[]): Map<string, SpiderfiedLocation> {
  const resultMap = new Map<string, SpiderfiedLocation>();
  const coordGroups = new Map<string, GeocodedRecord[]>();

  records.forEach(r => {
    if (r.lat === null || r.lon === null || r.precisa_revisao) return;
    const key = `${r.lat.toFixed(4)}|${r.lon.toFixed(4)}`;
    if (!coordGroups.has(key)) coordGroups.set(key, []);
    coordGroups.get(key)!.push(r);
  });

  coordGroups.forEach((group) => {
    const count = group.length;

    if (count === 1) {
      const item = group[0];
      resultMap.set(item.id, {
        id: item.id,
        originalLat: item.lat!,
        originalLon: item.lon!,
        displayLat: item.lat!,
        displayLon: item.lon!,
        isSpiderfied: false,
        spiderClusterCount: 1
      });
      return;
    }

    const radiusMeters = Math.min(35, 12 + count * 2.5);
    const latDelta = radiusMeters / 111000;
    const centerLat = group[0].lat!;
    const centerLon = group[0].lon!;
    const lonDelta = radiusMeters / (111000 * Math.cos((centerLat * Math.PI) / 180));

    group.forEach((item, idx) => {
      const angle = (2 * Math.PI * idx) / count;
      const displayLat = Number((centerLat + latDelta * Math.sin(angle)).toFixed(6));
      const displayLon = Number((centerLon + lonDelta * Math.cos(angle)).toFixed(6));

      resultMap.set(item.id, {
        id: item.id,
        originalLat: centerLat,
        originalLon: centerLon,
        displayLat,
        displayLon,
        isSpiderfied: true,
        spiderClusterCount: count
      });
    });
  });

  return resultMap;
}

export function exportToGeoJSON(records: GeocodedRecord[]): any {
  const spiderMap = applySpiderfyClustering(records);

  const features = records
    .filter(r => r.lat !== null && r.lon !== null && !r.precisa_revisao)
    .map(r => {
      const sp = spiderMap.get(r.id);
      return {
        type: 'Feature',
        geometry: {
          type: 'Point',
          coordinates: [sp ? sp.displayLon : r.lon, sp ? sp.displayLat : r.lat]
        },
        properties: {
          id: r.id,
          endereco_original: r.endereco_original,
          logradouro: r.logradouro,
          numero: r.numero,
          bairro: r.bairro,
          cep: r.cep,
          status_geocodificacao: r.status_geocodificacao,
          precisa_revisao: r.precisa_revisao,
          detalhe_resolucao: r.detalhe_resolucao,
          is_spiderfied: sp?.isSpiderfied || false,
          total_unidades_edificio: sp?.spiderClusterCount || 1,
          coord_original_lat: sp?.originalLat || r.lat,
          coord_original_lon: sp?.originalLon || r.lon
        }
      };
    });

  return {
    type: 'FeatureCollection',
    metadata: {
      total_analisados: records.length,
      geocodificados_sucesso: features.length,
      pendentes_revisao: records.filter(r => r.precisa_revisao).length,
      timestamp: new Date().toISOString()
    },
    features
  };
}

File: scripts/geocoding_pipeline.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
GEOSPATIAL DATA PIPELINE - ANTIGRAVITY ENGINE (PYTHON)
Arquitetura Determinística de Geocodificação & Saneamento Cadastral

Etapa 1: Parsing e Limpeza Estrita (Regex com Marcadores Explícitos)
Etapa 2: Cascata Hierárquica (Nível 1 Escrituras -> Nível 2 Geocode Predial -> Nível 3 Interpolação Via)
Etapa 3: Bloqueio Rigoroso de Falsa Precisão (Regra de Ouro)
Etapa 4: Preservação de Coordenadas Físicas Puras (Spiderfy Estritamente Visual no Frontend)
"""

import re
import json
import time
import unicodedata
import math
import urllib.parse
import urllib.request
import urllib.error
from typing import Dict, List, Optional, Tuple, Any

def normalize_str(s: Optional[str]) -> str:
    if not s:
        return ""
    nfkd = unicodedata.normalize('NFD', s)
    return "".join(c for c in nfkd if not unicodedata.combining(c)).strip()

# --------------------------------------------------------------------------
# CONTROLE DE TAXA NOMINATIM (1 req/segundo + Retry Exponencial)
# --------------------------------------------------------------------------
LAST_NOMINATIM_CALL_TIME = 0.0

def call_nominatim_with_rate_limit(url: str, headers: Dict[str, str], max_retries: int = 3) -> Optional[List[Dict[str, Any]]]:
    global LAST_NOMINATIM_CALL_TIME
    
    # Limite estrito de 1 requisição por segundo para proteger contra HTTP 429
    now = time.time()
    elapsed = now - LAST_NOMINATIM_CALL_TIME
    if elapsed < 1.05:
        time.sleep(1.05 - elapsed)
        
    backoff = 2.0
    for attempt in range(max_retries):
        try:
            LAST_NOMINATIM_CALL_TIME = time.time()
            req = urllib.request.Request(url, headers=headers)
            with urllib.request.urlopen(req, timeout=4.0) as resp:
                if resp.status == 200:
                    raw = resp.read().decode('utf-8')
                    data = json.loads(raw)
                    return data if isinstance(data, list) else [data]
        except urllib.error.HTTPError as e:
            # 429 = Too Many Requests, 503 = Service Unavailable
            if e.code in (429, 503):
                time.sleep(backoff)
                backoff *= 2.0
                continue
            else:
                break
        except Exception:
            time.sleep(1.0)
            continue
            
    return None

# --------------------------------------------------------------------------
# ETAPA 1: Parsing e Limpeza Estrita (Regex com Marcadores Explícitos)
# --------------------------------------------------------------------------
def parse_address_strict(
    raw_address: str,
    raw_neighborhood: str = "",
    raw_city: str = "",
    raw_state: str = "",
    raw_cep: str = ""
) -> Dict[str, Any]:
    text = normalize_str(raw_address)

    # 1. Extração Estrita de CEP: ^\d{5}-?\d{3}$
    cep = None
    cep_match = re.search(r'\b(\d{5})-?(\d{3})\b', raw_address)
    if not cep_match and raw_cep:
        cep_match = re.search(r'\b(\d{5})-?(\d{3})\b', raw_cep)
    if cep_match:
        cep = f"{cep_match.group(1)}-{cep_match.group(2)}"

    # Remove CEP do corpo do endereço
    text = re.sub(r'\b\d{5}-?\d{3}\b', '', text)
    text = re.sub(r'\bcep:?\b', '', text, flags=re.IGNORECASE).strip()

    # 2. Isolar Prefixo Padronizado (logradouro_tipo)
    prefix_regex = r'^(rua|r\.|r\b|avenida|av\.|av\b|estrada|estr\.|estr\b|travessa|trav\.|trav\b|alameda|al\.|al\b|praca|praça|pca\.|pca\b|rodovia|rod\.|rod\b|largo|lrg\.|lrg\b|beco|bc\.|bc\b|ladeira|lad\.|lad\b)'
    prefix_match = re.match(prefix_regex, text, flags=re.IGNORECASE)
    
    logradouro_tipo = "Rua"
    remaining = text

    if prefix_match:
        raw_p = prefix_match.group(1).lower().replace('.', '')
        if raw_p.startswith('av'):
            logradouro_tipo = 'Avenida'
        elif raw_p.startswith('estr'):
            logradouro_tipo = 'Estrada'
        elif raw_p.startswith('tra'):
            logradouro_tipo = 'Travessa'
        elif raw_p.startswith('al'):
            logradouro_tipo = 'Alameda'
        elif raw_p.startswith('pr') or raw_p.startswith('pc'):
            logradouro_tipo = 'Praca'
        elif raw_p.startswith('rod'):
            logradouro_tipo = 'Rodovia'
        elif raw_p.startswith('l'):
            logradouro_tipo = 'Largo'
        elif raw_p.startswith('bec') or raw_p.startswith('bc'):
            logradouro_tipo = 'Beco'
        elif raw_p.startswith('lad'):
            logradouro_tipo = 'Ladeira'
        else:
            logradouro_tipo = 'Rua'
        remaining = text[prefix_match.end():].strip()

    remaining = re.sub(r'^[-;,\.\s]+', '', remaining).strip()

    # 3. Número e Complemento com Marcador Estrito
    # Impede que vias como "Rua 24 de Maio, 120" ou "Av. 15 de Novembro, 450" tenham o número da via capturado
    numero = None
    complemento = ""
    logradouro_nome = remaining

    sn_regex = r'\b(s/n|sem numero|s/ numero|s\.n\.)\b'
    if re.search(sn_regex, remaining, flags=re.IGNORECASE):
        numero = None
        parts = re.split(sn_regex, remaining, flags=re.IGNORECASE)
        logradouro_nome = parts[0]
        complemento = " ".join(parts[1:])
    else:
        # Exigência de marcador explícito: vírgula OU palavras-chave (nº, n., numero, num, n)
        # Ignora números isolados que façam parte do nome do logradouro
        num_pattern = re.compile(
            r'^(.*?)(?:,\s*(?:n[ºo°\.]*|n\b|num\b|numero\b)?\s*|[\s]+(?:n[ºo°\.]+|n\.\s*|num\b|numero\b)\s*:?\s*)(\d{1,6})\b(.*)$',
            re.IGNORECASE
        )
        num_match = num_pattern.match(remaining)
        if num_match:
            logradouro_nome = num_match.group(1)
            numero = num_match.group(2)
            complemento = num_match.group(3) or ""
        else:
            comp_regex = r'\b(apto|apt|apartamento|ap|bl|bloco|qd|quadra|lt|lote|cs|casa|fundos|fdo|sobrado|sob|sala|unidade|un|pavimento|pav|andar|edificio|edf|condominio)\b'
            comp_match = re.search(comp_regex, remaining, flags=re.IGNORECASE)
            if comp_match:
                logradouro_nome = remaining[:comp_match.start()]
                complemento = remaining[comp_match.start():]

    # Limpeza de Logradouro Nome
    logradouro_nome = re.sub(r'[,\-\.]+', ' ', logradouro_nome)
    logradouro_nome = re.sub(r'\s+', ' ', logradouro_nome).strip()
    logradouro_nome = " ".join(w.capitalize() for w in logradouro_nome.split())

    # Limpeza de Complemento
    complemento = re.sub(r'^[-;,\.\s:]+|[-;,\.\s:]+$', '', complemento)
    complemento = re.sub(r'\s+', ' ', complemento).strip().upper()

    # Bairro, Cidade, UF
    bairro = re.sub(r'[^A-Z0-9\s]', '', normalize_str(raw_neighborhood).upper()).strip()
    cidade = re.sub(r'[^a-zA-Z0-9\s]', '', normalize_str(raw_city)).strip() if raw_city else "Rio de Janeiro"
    uf = raw_state.strip().upper() if raw_state else "RJ"

    return {
        "logradouro_tipo": logradouro_tipo,
        "logradouro_nome": logradouro_nome,
        "numero": numero,
        "complemento": complemento,
        "bairro": bairro,
        "cidade": cidade,
        "uf": uf,
        "cep": cep
    }

# --------------------------------------------------------------------------
# ETAPA 2 & 3: Cascata e Bloqueio de Falsa Precisão
# --------------------------------------------------------------------------
def resolve_coordinates_cascade(
    record_id: str,
    raw_address: str,
    raw_neighborhood: str = "",
    raw_city: str = "",
    raw_state: str = "",
    raw_cep: str = "",
    internal_cache: Optional[Dict[str, Tuple[float, float]]] = None
) -> Dict[str, Any]:
    parsed = parse_address_strict(raw_address, raw_neighborhood, raw_city, raw_state, raw_cep)
    full_logradouro = f"{parsed['logradouro_tipo']} {parsed['logradouro_nome']}".strip()

    # NÍVEL 1: Base Própria de Escrituras / IPTU / ITBI (EXATO_PREDIO)
    if parsed["numero"] and internal_cache:
        cache_key = f"{parsed['logradouro_nome'].lower()}|{parsed['numero']}|{parsed['bairro'].lower()}"
        if cache_key in internal_cache:
            coords = internal_cache[cache_key]
            return {
                "id": record_id,
                "endereco_original": raw_address,
                "logradouro": full_logradouro,
                "numero": parsed["numero"],
                "bairro": parsed["bairro"],
                "cep": parsed["cep"],
                "lat": round(coords[0], 6),
                "lon": round(coords[1], 6),
                "status_geocodificacao": "EXATO_PREDIO",
                "precisa_revisao": False,
                "detalhe_resolucao": "Base Oficial de Escrituras/ITBI Municipal"
            }

    # NÍVEL 2 & NÍVEL 3: Geocodificação Estruturada via API Nominatim com Rate Limit
    try:
        street_param = f"{parsed['numero']} {parsed['logradouro_tipo']} {parsed['logradouro_nome']}" if parsed["numero"] else full_logradouro
        params = {
            "street": street_param,
            "city": parsed["cidade"],
            "state": parsed["uf"],
            "country": "Brazil",
            "format": "jsonv2",
            "addressdetails": "1"
        }
        if parsed["cep"]:
            params["postalcode"] = parsed["cep"]

        query_str = urllib.parse.urlencode(params)
        url = f"https://nominatim.openstreetmap.org/search?{query_str}"
        headers = {'User-Agent': 'MarcusAssessoriaGarimpo/2.0 (Deterministic Geocoding Engine)'}
        
        data = call_nominatim_with_rate_limit(url, headers=headers)
        if data and len(data) > 0:
            top = data[0]
            lat = float(top.get("lat", 0))
            lon = float(top.get("lon", 0))
            place_rank = top.get("place_rank", 0)
            category = top.get("category", "")
            type_ = top.get("type", "")

            # NÍVEL 2: Nível Predial / Lote / Número (house, building, address, place_rank >= 30)
            if place_rank >= 30 or type_ in ['house', 'building'] or category == 'building':
                return {
                    "id": record_id,
                    "endereco_original": raw_address,
                    "logradouro": full_logradouro,
                    "numero": parsed["numero"],
                    "bairro": parsed["bairro"],
                    "cep": parsed["cep"] or top.get("address", {}).get("postcode"),
                    "lat": round(lat, 6),
                    "lon": round(lon, 6),
                    "status_geocodificacao": "GEOCODE_NUMERO",
                    "precisa_revisao": False,
                    "detalhe_resolucao": f"Nominatim Predial (place_rank {place_rank}, {type_})"
                }

            # NÍVEL 3: Interpolação de Via / Eixo de Rua (highway, residential, road, place_rank 26-28)
            if (26 <= place_rank <= 28) or category == "highway":
                return {
                    "id": record_id,
                    "endereco_original": raw_address,
                    "logradouro": full_logradouro,
                    "numero": parsed["numero"],
                    "bairro": parsed["bairro"],
                    "cep": parsed["cep"] or top.get("address", {}).get("postcode"),
                    "lat": round(lat, 6),
                    "lon": round(lon, 6),
                    "status_geocodificacao": "INTERPOLACAO_RUA",
                    "precisa_revisao": False,
                    "detalhe_resolucao": f"Interpolação no Eixo da Via (place_rank {place_rank})"
                }
    except Exception:
        pass

    # ETAPA 3: Bloqueio Rigoroso de Falsa Precisão (Regra de Ouro)
    # JAMAIS atribuir coordenadas de centróide de bairro ou cidade!
    return {
        "id": record_id,
        "endereco_original": raw_address,
        "logradouro": full_logradouro,
        "numero": parsed["numero"],
        "bairro": parsed["bairro"],
        "cep": parsed["cep"],
        "lat": None,
        "lon": None,
        "status_geocodificacao": "PENDENTE_REVISAO",
        "precisa_revisao": True,
        "detalhe_resolucao": "Logradouro não validado com certeza cartográfica predial ou viária"
    }

# --------------------------------------------------------------------------
# ETAPA 4: GeoJSON com Coordenadas Puras (Sem Contaminação Espacial)
# --------------------------------------------------------------------------
def export_to_geojson(records: List[Dict[str, Any]]) -> Dict[str, Any]:
    features = []

    # Contagem de unidades por coordenada física para alimentar metadados visuais
    coord_counts: Dict[str, int] = {}
    for r in records:
        if r["lat"] is not None and r["lon"] is not None and not r["precisa_revisao"]:
            k = f"{r['lat']:.4f}|{r['lon']:.4f}"
            coord_counts[k] = coord_counts.get(k, 0) + 1

    for r in records:
        if r["lat"] is None or r["lon"] is None or r["precisa_revisao"]:
            continue
            
        k = f"{r['lat']:.4f}|{r['lon']:.4f}"
        cluster_count = coord_counts.get(k, 1)

        # As coordenadas físicas REAIS permanecem 100% puras [lon, lat]!
        # Spiderfy é estritamente uma camada de visualização no Leaflet (frontend).
        features.append({
            "type": "Feature",
            "geometry": {
                "type": "Point",
                "coordinates": [r["lon"], r["lat"]]
            },
            "properties": {
                "id": r["id"],
                "endereco_original": r["endereco_original"],
                "logradouro": r["logradouro"],
                "numero": r["numero"],
                "bairro": r["bairro"],
                "cep": r["cep"],
                "status_geocodificacao": r["status_geocodificacao"],
                "precisa_revisao": r["precisa_revisao"],
                "detalhe_resolucao": r.get("detalhe_resolucao"),
                "total_unidades_edificio": cluster_count
            }
        })

    return {
        "type": "FeatureCollection",
        "metadata": {
            "total_analisados": len(records),
            "geocodificados_sucesso": len(features),
            "pendentes_revisao": sum(1 for r in records if r["precisa_revisao"])
        },
        "features": features
    }

if __name__ == "__main__":
    test_cases = [
        ("1", "RUA 24 DE MAIO, 120 APTO 101", "Riachuelo", "Rio de Janeiro", "RJ", "20950-090"),
        ("2", "AV. 15 DE NOVEMBRO, Nº 450", "Centro", "Petrópolis", "RJ", "25685-050"),
        ("3", "RUA ADAIL, N. 266 APTO 606 BL 02 QD 14 LT 08", "Penha Circular", "Rio de Janeiro", "RJ", "21210-000"),
        ("4", "RUA ENGENHEIRO ARTUR MOURA, N. 456, Apto 311, BL 4", "Bonsucesso", "Rio de Janeiro", "RJ", ""),
        ("5", "RUA FANTASMA INEXISTENTE ZZZ, 9999", "Bonsucesso", "Rio de Janeiro", "RJ", "")
    ]

    print("--- Executando Testes do Pipeline Determinístico ---")
    results = []
    for t in test_cases:
        p = parse_address_strict(t[1], t[2], t[3], t[4], t[5])
        print(f"Parsed: '{t[1]}' -> Logradouro: '{p['logradouro_tipo']} {p['logradouro_nome']}' | Num: '{p['numero']}' | Comp: '{p['complemento']}'")
        res = resolve_coordinates_cascade(t[0], t[1], t[2], t[3], t[4], t[5])
        results.append(res)
        print(f"  => [{res['status_geocodificacao']}] Lat: {res['lat']}, Lon: {res['lon']} (Revisao: {res['precisa_revisao']})")

    geojson_out = export_to_geojson(results)
    print("\nGeoJSON Feature Count:", len(geojson_out["features"]))
    print("Metadata:", json.dumps(geojson_out["metadata"], indent=2))
